### PR TITLE
Bugfix/5303 reroute yflow correct response if already on best path

### LINKED ISSUE #5303: GUI/Kilda: Reroute y-flow button cause the wrong message in case if we already on the best flow.
Right now if we press the reroute y-flow button for the one of the Y-Flow's subflow it will show us the following popup window in the top right corner and also does not show us the flow path in the bottom part of the screen: 
image
The problem is that the Kilda northbound API returns the 500 response on calling this reroute API:
`http://172.19.123.246:8080/api/v2/y-flows/yflow8-9-3/reroute`
here is the response: 
image

In fact, it would not let us reroute because we already on the best route for the current Y-Flow and the 500 message that we have does not show us the actual reason, for example for the similar case, but just the regular flow, we have the following response, if we already on the best route:
image
and for the UI part it shows the following message:
image

We need to make the same behaviour for the Y-Flow. 
- [ ] Kilda: fix the northbound API response, fix FlowHs topology to correctly handle the case with the "already on the best route"
- [ ] UI: handle correctly "already on the best route" case. 

PRs:
https://github.com/telstra/open-kilda/pull/5318
https://github.com/telstra/open-kilda/pull/5293

### PR DESCRIPTION
**Pull request by IvanChupin**
Added the correct response for Y-Flow reroute API, for the case of "already on the best path".
Slightly changed the logic for Y-flow-reroute functionality for partly failure cases.

Here is the picture that represents the 6 cases covered in this fix.
Basically we have y-flow with the 9th switch as a y-point and shared switch.
3rd switch is a first endpoint, 8th switch as a second endpoint.
In the picture, for cases 1,2,3 and 4 we can see an existed possible path between switches that has less cost than the current one.
For the 3,4,5,6 cases we can see that there is no possible and any other paths with one of the endpoints.

image
From the picture above you can see that for only 5th case was a correct reroute response with appropriate behaviour. 
This PR fixes the rest of the cases.

What has been change in terms of a internal logic?

**BEFORE**: Y-flow reroute fsm launches the regular reroute fsm under the hood for each subflow. the problem was that if we face any error while rerouting the first subflow we will not try to reroute the second subflow, and as result we will fail the whole y-flow reroute process. 
Here is the description for each 6 case of an old behaviour:
Y-flow 9-3-8 ,     affinityGroup subflow9-3 

**1.** There is an alternate subpath 9-3, for 9-8 the same.  Response 500, ﻿ but reroute  9-3 happened.   This is wrong behaviour. The response body : 
```
{.  
  "correlation_id": "25fe7ce5-0254-4a8d-bcf3-1d61f6c2cce9 : reruote9-3-8",
  "timestamp": 1691066388920,
  "error-type": "Internal service error",
  "error-message": "Could not reroute y-flow",
  "error-description": "Reroute is unsuccessful. Couldn't find new path(s)"
}
```
**2.** There is an alternate subpath 9-8, for 9-3 the same.  Response 500. Reroute has not been happened . This is wrong behaviour.
**3.** There is an alternate subpath  9-3, for 9-8 no path, since 8 is unreachable(switch 8  all isl ports down [1, 7, 16, 47, 49]): X Got 404, but  9-3 has not been rerouted, This is wrong behaviour. The response body: 
```
 {
  "correlation_id": "14788401-8ab9-424a-842a-6b1714c4cfee : reruote9-3-8",
  "timestamp": 1691067035992,
  "error-type": "Object was not found",
  "error-message": "Could not reroute y-flow",
  "error-description": "Not enough bandwidth or no path found. Switch 00:00:00:00:00:00:00:08 doesn't have links with enough bandwidth, Failed to find path with requested bandwidth=100"
}
```
**4.** There is an alternate subpath 9-8, for 9-3 no path, since 3 is unreachable (switch 3 all isl ports down [1,2,4,5,6] ) :  Got 404:   nothing was rerouted. This is wrong behaviour. The response body:     
```
{
  "correlation_id": "6799435d-9c4b-4a03-b557-f288733c464d : reruote9-3-8",
  "timestamp": 1691068142508,
  "error-type": "Object was not found",
  "error-message": "Could not reroute y-flow",
  "error-description": "Not enough bandwidth or no path found. Switch 00:00:00:00:00:00:00:03 doesn't have links with enough bandwidth, Failed to find path with requested bandwidth=100"
}
```
**5.** No alternate paths for 9-8.  9-3 impossible, no path.  since 3 is unreachable.: Response 404:   Correct response with the correct response body:     
```
{
  "correlation_id": "59b9c4df-884f-48f5-9240-38f61ecfe466 : reruote9-3-8",
  "timestamp": 1691068808567,
  "error-type": "Object was not found",
  "error-message": "Could not reroute y-flow",
  "error-description": "Not enough bandwidth or no path found. Switch 00:00:00:00:00:00:00:03 doesn't have links with enough bandwidth, Failed to find path with requested bandwidth=100"
}
```
**6.** No alternate paths for 9-3, 9-8 impossible, no path. 8 unreachable. Response 500, Wrong response. Response body:     
```
 {
  "correlation_id": "5be8981a-aa81-44f0-a85a-94f02627848b : reruote9-3-8",
  "timestamp": 1691069310602,
  "error-type": "Internal service error",
  "error-message": "Could not reroute y-flow",
  "error-description": "Reroute is unsuccessful. Couldn't find new path(s)"
}
```

**AFTER**: 
+ Y-flow reroute fsm will not stop the attempt to reroute the second subflow in case of a first subflow reroute failure. 
+ added correct response bodies, replace 500 response code with 404.

To reproduce the behaviour we could use the following req:
`http://vm:8080/api/v2/y-flows`

```
{
  "y_flow_id": "yflow9-3-8",
  "shared_endpoint": {
    "port_number": 101,
    "switch_id": "9"
  },
  "maximum_bandwidth": 100,
  "path_computation_strategy": "cost",
  "sub_flows": [
    {
      "flow_id": "subflow9-3",
      "endpoint": {
        "detect_connected_devices": {
          "arp": true,
          "lldp": true
        },
        "inner_vlan_id": 108,
        "port_number": 103,
        "switch_id": "3",
        "vlan_id": 109
      },
      "shared_endpoint": {
        "inner_vlan_id": 110,
        "vlan_id": 111
      }
    },
{
      "flow_id": "subflow9-8",
      "endpoint": {
        "detect_connected_devices": {
          "arp": true,
          "lldp": true
        },
        "inner_vlan_id": 118,
        "port_number": 113,
        "switch_id": "8",
        "vlan_id": 120
      },
      "shared_endpoint": {
        "inner_vlan_id": 111,
        "vlan_id": 112
      }
    }
  ]
}
```

then we can alter the cost for the isl links between certain switches to reproduce 6 scenarios from the screenshot(or from the description list)

closes #5303

### CODE REVIEW

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/error/ErrorType.java [hunk 1]
```diff
@@ -101,7 +101,12 @@ public enum ErrorType {
     /**
      * The request cannot be processed.
      */
-    UNPROCESSABLE_REQUEST("The request cannot be processed");
+    UNPROCESSABLE_REQUEST("The request cannot be processed"),
+
+    /**
+     * The error message for Y-Flow subflow that can not be rerouted since already on the best path.
+     */
+    ALREADY_ON_BEST_PATH("The sub-flow already on the best path");
```
Comment (niksv) [inline src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/error/ErrorType.java:109]: NOTE: this list of error is very generic, but this error description is very specific. 
I mean `ALREADY_ON_BEST_PATH` is good error, but `"The sub-flow already on the best path"` is only about sub flows, however `ALREADY_ON_BEST_PATH` could also be used by common flows. Maybe it's better to change description to `The flow already on the best path`?

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/actions/PostResourceAllocationAction.java [hunk 2]
```diff
@@ -62,10 +62,12 @@ protected Optional<Message> performWithResponse(State from, State to, Event even
             currentForwardId = flow.getForwardPathId();
         }
         FlowPath currentForwardPath = currentForwardId != null ? getFlowPath(currentForwardId) : null;
-
+        Message rerouteResponse = buildRerouteResponseMessage(currentForwardPath, newForwardPath,
+                stateMachine.getCommandContext());
         if (stateMachine.getNewPrimaryForwardPath() == null && stateMachine.getNewPrimaryReversePath() == null
                 && stateMachine.getNewProtectedForwardPath() == null
                 && stateMachine.getNewProtectedReversePath() == null) {
+            stateMachine.setOperationResultMessage(rerouteResponse);
```
Comment (niksv) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/actions/PostResourceAllocationAction.java:70]: What will happen if we would find new path, because it doesn't exists. I mean current path is broken and there is no other path. Will we return "Flow already on the best path"?

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/error/ErrorType.java (hunk 1 above)
Comment (IvanChupin) [reply src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/error/ErrorType.java]: Thanks, fixed!

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/actions/PostResourceAllocationAction.java (hunk 2 above)
Comment (IvanChupin) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/actions/PostResourceAllocationAction.java:70]: The following message:

{
  "correlation_id": "57501d77-f954-4771-8cd1-b5e126271cc7 : sdasd",
  "timestamp": 1690970708697,
  "error-type": "Object was not found",
  "error-message": "Could not reroute y-flow",
  "error-description": "Not enough bandwidth or no path found. Switch 00:00:00:00:00:00:00:08 doesn't have links with enough bandwidth, Failed to find path with requested bandwidth=100"
}

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/FlowRerouteFsm.java [hunk 3]
```diff
@@ -466,6 +464,19 @@ public FlowRerouteFsm newInstance(@NonNull String flowId, @NonNull CommandContex
             });
             return fsm;
         }
+
+        private ErrorType extractErrorType(FlowRerouteFsm fsm) {
+            return Optional.ofNullable(fsm.getOperationResultMessage())
+                    .filter(message -> message instanceof ErrorMessage || message instanceof InfoMessage)
+                    .map(message -> {
+                        if (message instanceof InfoMessage) {
+                            return ErrorType.ALREADY_ON_BEST_PATH;
```
Comment (pablomuri) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/FlowRerouteFsm.java:473]: Is this tested?

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/yflow/reroute/actions/HandleNotReroutedSubFlowAction.java [hunk 4]
```diff
@@ -76,14 +86,31 @@ protected Optional<Message> performWithResponse(State from, State to, Event even
...
         if (isFirstError) {
-            Message message = stateMachine.buildErrorMessage(context.getErrorType(), getGenericErrorMessage(),
-                    context.getError());
+            Message message;
+            if (ErrorType.ALREADY_ON_BEST_PATH.equals(context.getErrorType())) {
+                message = buildRerouteResponseMessage(stateMachine);
+            } else {
+                message = stateMachine.buildErrorMessage(context.getErrorType(), getGenericErrorMessage(),
+                        context.getError());
+            }
```
Comment (pablomuri) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/yflow/reroute/actions/HandleNotReroutedSubFlowAction.java:95]: Is this tested?

**Review by pablomuri — CHANGES_REQUESTED** [review]

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/FlowRerouteFsm.java (hunk 3 above)
Comment (IvanChupin) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/FlowRerouteFsm.java:473]: thanks
cover with the unit tests all the corner cases

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/yflow/reroute/actions/HandleNotReroutedSubFlowAction.java (hunk 4 above)
Comment (IvanChupin) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/yflow/reroute/actions/HandleNotReroutedSubFlowAction.java]: thanks
cover with the unit tests all the corner cases

**Review by pablomuri — APPROVED** [review]
great job!

**Review by niksv — APPROVED** [review]

**Review by dmitrii-beliakov — APPROVED** [review]
